feat: Implement `n_distinct()` for multiple arguments using duckdb structs

lschneiderbauer opened the pull request:
This PR attempts to adress #110, implementing `n_distinct()` making use of duckdb's [structs](https://duckdb.org/docs/sql/data_types/struct).

I don't have any experience in writing dbplyr SQL translations, in particular I have no idea of the 'best practices', but I tried to mimic as much from the already existing code as I could.

A couple of notes:

- duckdb's struct fields need to be named. In this case, the naming is irrelevant and I chose to name them 'v1', 'v2', 'v3', ... and so on, depending on the number of arguments passed to `n_distinct()`. If there is an alternative prefered way to do it, please let me know.
- I was making use of `glue::glue()`. Although dbplyr depends on glue, and in this function we are assuming that dbplyr is installed, I am not sure if I am allowed to assume that this function exists. Maybe I should not make use of it?
- I also added unit tests at places where I thought they make sense. Let me know if I should change them or relocate them, etc..

Thanks! :)

Closes #110

krlmlr commented on R/backend-dbplyr__duckdb_connection.R:
```diff
@@ -77,6 +77,26 @@ duckdb_grepl <- function(pattern, x, ignore.case = FALSE, perl = FALSE, fixed =
 }
 
+duckdb_n_distinct <- function(..., na.rm = FALSE) {
+  check_na_rm <- pkg_method("check_na_rm", "dbplyr")
+  glue_sql2 <- pkg_method("glue_sql2", "dbplyr")
```
Those are private functions. @mgirlich: Is it fair to assume that these will stay around?

mgirlich replied:
`check_na_rm()` should definitely stay and might be worth exporting for backends. I opened an issue: https://github.com/tidyverse/dbplyr/issues/1483.
It would also make sense to export `glue_sql2()` but I don' think this is ready yet. So, I guess this is more likely to evolve (in particular I would consider changing the name before exporting it).

krlmlr reviewed (commented):
Thanks! A small integration test (what does `n_distinct()` actually do with a real dummy table) would be useful too.

Sorry about the conflicts.

lschneiderbauer commented:
> A small integration test (what does `n_distinct()` actually do with a real dummy table) would be useful too.
I added a bunch of tests to check the actual computation.

In the meantime I also realized that my assumption that duckdb would use a `na.rm = TRUE` semantics in this case is actually wrong: NULLs within structs seem to actually be counted as a single unique entity. The added test results suggest that.
So I removed the wrong `check_na_rm()` and opted to throw an error if the `na.rm = TRUE` is passed. Since `na.rm = FALSE` is the default this seemed appropriate.
If one wanted to implemented a `na.rm = TRUE` behavior, using the [`FILTER`](https://duckdb.org/docs/sql/query_syntax/filter) clause might be a way to go, but according to the documentation this does not work within a windowing context. Also, it's not obvious to me how to check any element in a struct for NULLs with the [struct functions](https://duckdb.org/docs/sql/functions/nested.html#struct-functions) that are available so far.

Regarding the use of `dbplyr:::glue_sql2()`, what should be the way to go: Wait with the PR until such a function becomes exported, reimplement it in this package, or keep using the private function?

krlmlr commented:
Thanks. If we must, we can reimplement here, or use plain `paste0()` ?

lschneiderbauer commented:
Thanks.

- I am using the `dbplyr::sql()` function directly now, together with `paste0()`, instead of `dbplyr:::glue_sql2()`.
- `cli::cli_abort()` is replaced by `stop()` to get rid of R CMD CHECK warning.
- I am making use of duckdb's [row()](https://duckdb.org/docs/sql/data_types/struct.html#creating-structs-with-the-row-function) function now instead of using the '{ ... }' syntax which makes generating SQL easier. (and also gets rid of `glue::glue()` to make R CMD CHECK happy)

krlmlr commented on R/backend-dbplyr__duckdb_connection.R:
```diff
@@ -77,6 +77,18 @@ duckdb_grepl <- function(pattern, x, ignore.case = FALSE, perl = FALSE, fixed =
+duckdb_n_distinct <- function(..., na.rm = FALSE) {
+  sql <- pkg_method("sql", "dbplyr")
+
+  if (identical(na.rm, TRUE)) {
```
Slightly stricter:

```suggestion
  if (!identical(na.rm, FALSE)) {
```

lschneiderbauer commented:
@krlmlr Anything else you would like me to improve?

krlmlr commented on R/backend-dbplyr__duckdb_connection.R:
```diff
@@ -77,6 +77,18 @@ duckdb_grepl <- function(pattern, x, ignore.case = FALSE, perl = FALSE, fixed =
+  }
+
+  # https://duckdb.org/docs/sql/data_types/struct.html#creating-structs-with-the-row-function
+  str_struct <- paste0("row(", paste0(list(...), collapse = ", "), ")")
+
+  sql(paste0("COUNT(DISTINCT ", str_struct, ")"))
```
@mgirlich: Can you please help me review this part of the code?

mgirlich replied:
Basically looks good to me. This also works for diverse user input, e.g. `n_distinct(x, a / (b - 1))` or `n_distinct(!!x)`.
One minor thing: you might want to check for empty dots.

krlmlr reviewed (commented):
Thanks for the nudge!

mgirlich commented on tests/testthat/test-backend-dbplyr__duckdb_connection.R:
```diff
@@ -218,8 +221,54 @@ test_that("two variable aggregates are translated correctly", {
+    pull(summarize(df_na, n = n_distinct(x, y)), n),
+    5
+  )
+})
```
You might want to add a test that uses a window.

krlmlr commented:
Thanks! Adding the extra tests and checks could be done in a small follow-up PR.
